### PR TITLE
Programmentwurf chapter on unit tests

### PR DESCRIPTION
**Pull request by erikzmmrmnn**

### CODE REVIEW

File: LaTeX/Inhalt/04_Inhalt/unit_tests.tex [hunk 1]
```diff
@@ -1 +1,25 @@
-\chapter{Unit Tests}
\ No newline at end of file
+\chapter{Unit Tests}
+Die Unit Tests von Morik wurden so geschrieben, dass sie uns beim Entwickeln neuer Funktionalitäten helfen. Als Test-Framework wird dabei \href{https://github.com/moorts/Morik/blob/main/src/CMakeLists.txt#L24}{\textit{GoogleTest}} verwendet. Um eine gute Übersicht über die Tests zu haben, sind diese genau in der gleichen Ordnerstruktur angeordnet wie der Produktivcode, nur eben im \href{https://github.com/moorts/Morik/tree/main/src/tests}{\textit{tests}}-Ordner. Im Folgenden wird genauer auf die Umsetzung der ATRIP-Regeln, das Messen der Code Coverage und den Einsatz von Mock-Objekten eingegangen.
+
+\section{ATRIP-Regeln}
```
Comment (moorts) [inline LaTeX/Inhalt/04_Inhalt/unit_tests.tex:4]: Maybe add section description (short text explaining what ATRIP is and such

File: LaTeX/Inhalt/04_Inhalt/unit_tests.tex [hunk 2]
```diff
@@ -1 +1,25 @@
...
...n zu können, ohne vorher alle Abhängigkeiten konstruieren zu müssen, werden an ausgewählten Stellen Mock-Objekte eingesetzt. Für das Erzeugen der Mock-Objekte wird \textit{gmock} verwendet, das in GoogleTest enthalten ist. So kann in der Klasse \href{https://github.com/moorts/Morik/blob/main/src/tests/application/AbstractDatabaseInterfaceMock.h}{\textit{AbstractDatabaseInterfaceMock}} bspw. das Mocken eines AbstractDatabaseInterface beobachtet werden, sodass in \href{https://github.com/moorts/Morik/blob/main/src/tests/application/repository_tests.cpp}{\textit{repository\_tests}} die EntryRepository getestet werden kann, die die Übergabe eines AbstractDatabaseInterface im Konstruktor erwartet. Ein weiteres Mock-Objekt ist das \href{https://github.com/moorts/Morik/blob/main/src/tests/adapters/database/AbstractSqlDatabaseMock.h}{\textit{AbstractSqlDatabaseMock}}, das eine AbstractSqlDatabase mockt. Dies wird verwendet, um in \href{https://github.com/moorts/Morik/blob/main/src/tests/adapters/database/databaseInterface_tests.cpp}{\textit{databaseInterface\_tests}} den DatabaseInterface-Adapter testen zu können, der wiederum die Übergabe einer AbstractSqlDatabase im Konstruktor erwartet.
```
Comment (moorts) [inline LaTeX/Inhalt/04_Inhalt/unit_tests.tex:25]: Perhaps explain broadly how mock classes work (like how return values are manually set and such) and how the mock classes are configured in the tests...

This should be brief though, if you include it at all

File: LaTeX/Inhalt/04_Inhalt/unit_tests.tex [hunk 3]
```diff
@@ -1 +1,25 @@
...
+
+\subsection{Professional}
+Der Testcode von Morik ist professionell geschrieben. Dies bedeutet, dass er leicht verständlich ist aufgrund von sprechenden Bezeichnern und dem Befolgen der AAA-Normalform. Außerdem werden keine unrelevanten Aspekte der Applikation getestet, sondern wir fokussieren uns, wie bereits bei ''Thorough'' genannt, auf das Testen der wichtigsten Funktionalitäten.
+
+\section{Code Coverage}
+Wie zuvor bereits genannt, ermöglicht die Code Coverage eine Einschätzung wie ''gründlich'' die Tests tatsächlich sind, indem gemessen wird wie viel Code beim Testen durchlaufen wird im Vergleich zu wie viel Code insgesamt in der kompletten Applikation durchlaufen werden kann. Hierbei werden zwei Metriken besonders häufig verwendet. Zum einen gibt es die Line Coverage, die die ausgeführten Codezeilen zählt, während es zum anderen die Branch Coverage gibt. Diese zählt die ausgeführten Verzweigungsmöglichkeiten. Dabei ist letztere Metrik aussagekräftiger für die tatsächliche ''Gründlichkeit'' der Tests, während die Line Coverage allerdings leichter zu messen ist.\\
```
Comment (moorts) [inline LaTeX/Inhalt/04_Inhalt/unit_tests.tex:21]: Is this accurate? I guess when a program has many early returns, branch coverage will differ greatly, but as can be seen in our case, the metrics are extremely similar. Also, in certain programs, branch coverage could even be misleading if most branches are barely called.

My first instinct is also to say that branch coverage is more accurate, but maybe we should cite literature or something to support that claim.

**Review by moorts — CHANGES_REQUESTED** [review]

File: LaTeX/Inhalt/04_Inhalt/unit_tests.tex (hunk 2 above)
Comment (erikzmmrmnn) [reply LaTeX/Inhalt/04_Inhalt/unit_tests.tex:25]: I think we can see the matters explained to us in the lecture as given knowledge. I would therefore assume that the reader knows how mock objects work.

File: LaTeX/Inhalt/04_Inhalt/unit_tests.tex (hunk 3 above)
Comment (erikzmmrmnn) [reply LaTeX/Inhalt/04_Inhalt/unit_tests.tex:21]: Lindner said that in his lecture, which is why I didn't back that claim by any source. Also if a program has many early returns those would probably be in a branch themselves which means that this branch as well as the other branch without the early return would need execution for coverage to reach 100%, whereas line coverage can go very high without ever executing a specific branch if that branch has only few lines. Therefore branch coverage is more accurate.

**Review by moorts — APPROVED** [review]

**Pull request merged**